### ISSUE TITLE
Optimisation: Save a byte in popa.s (most trivial optimisation ever?)

### ISSUE THREAD

**Issue by prandeamus** [cc65/cc65#566]
runtime/popa.s ends with this sequence, to do a 16-bit addition to SP

    inc     sp
    beq     @L1
    rts
@L1:    inc     sp+1
    rts

It would be a byte shorter if coded as:

    inc sp
    bne @L1
    inc sp+1
@L1: rts

This also makes the code consistent with incsp1.s

The number of cycles will be the same although by removing the extra RTS it reduces the chance that the branch will cross a page boundary, so maybe you save on average 1/256 of a cycle per call. :)

I was quite excited when I first saw this, then discovered that popa isn't ever called from the compiler, although it is called from some of the support libraries. Hey, if anyone wants to save a byte in one of those constrained environments, feel free to make this change. It's still the most productive I'll be today. :)

**Comment by spiro-trikaliotis** [comment]
The question is: Is there a non-trivial runtime penalty of your change?

The original code takes the branch "BEQ" only in 1 of 256 cases (if SP == 0 after the INC). The modified code takes the branch "BNE" in 255 of 256 cases.

As this is rather highly-used code, I am not sure if this approach was not done with a purpose, to prevent too many spendings of 1 clock cycle.

**Comment by prandeamus** [comment]
Interesting point. In that case, would there be a corresponding case to keep this exactly as it is, but modify incsp1.s to match it? (I also realise this is like discussing how many angels can dance on the head of a pin, and it may not be worth making such a change. It's fun, though!).

**Comment by oliverschmidt** [comment]
If there's at some point a common understanding which code variant is in general preferable I'm of course interested in a pull request harmonizing the runtime code to "always" use that variant.

**Comment by greg-king5** [comment]
Uz put cycle counts in the comments,  It tells us that he thought that that particular function's speed is more important than its size.

Look at "incsp2.s".  That same merge should be done with "popa.s" and "incsp1.s".

**Comment by mrdudz** [comment]
yeah this kind of functions are called an awful lot - speed is more important than size here.

that said, we could have two versions of the runtime - one optimized for speed and one optimized for size.

**Comment by oliverschmidt** [comment]
> we could have two versions of the runtime

Please don't.

**Comment by mrdudz** [comment]
its quite a logical thing to do, IMHO. also a runtime that uses the so called illegal opcodes would be quite useful (eg lax #0 comes to mind)

**Comment by prandeamus** [comment]
You could perhaps have special versions for undocumented opcode tricks that but i don't think it should be part of the official cc65 project . Maintenence would be a nightmare. Within an individual project you could write alternate versions  of runtime routines and persuade the linker to use them. 
There might be an argument for two runtime variants, one smaller and one faster, much as we have different -O compiler options. There will never be one setting that fits everyone especially given the limits of 6502 architecture. We have to accept the subroutines are essential to have sane code sizes. 
I would be interested in a code review of what we have. See if there are any optimisations on common routines that would benefit all users. That's what i was exploring with this discussion on  popa, my earlier question on leasp, and the discussion branch on the booth multiply algorithm.
Maybe after that, with a sound understanding of the runtime we can look at tweaking thecode generator, in ways that benefit all.

Sent from Yahoo Mail on Android 
 
  On Sat, 6 Jan 2018 at 13:05, Bob Andrews<notifications@github.com> wrote:   
its quite a logical thing to do, IMHO. also a runtime that uses the so called illegal opcodes would be quite useful (eg lax #0 comes to mind)

**Comment by mrdudz** [comment]
> You could perhaps have special versions for undocumented opcode tricks that but i don't think it should be part of the official cc65 project . Maintenence would be a nightmare.

why? i'd add it as a seperate CPU type, then also the code generator could take advantage of it

**Comment by spiro-trikaliotis** [comment]
While discussing all these changes, the question is: Has this ever been benchmarked? Yes, I argued with the extra cycle in many cases above, @greg-king5 pointed out incsp2 (which I had seen myself, too).
However, I am not quite sure how often these functions are actually used (that is, not how many static calls we have, but how often they are used dynamically - that can be different!), and if it is really worth the effort.

Are we speaking about 1 cycle in 1.000.000 cycles, 1 cycle in 10.000 cycles, or 1 cycle in 50 cycles here?

Does anyone know of some program (non-trivial, but deterministic) that could be used for such a benchmark?

**Comment by prandeamus** [comment]
I was thinking about extending the paravirtualisaton hooks in sim6502 so we could show cycle time changes with more precision on fragments of code. 

Sent from Yahoo Mail on Android 
 
  On Sat, 6 Jan 2018 at 14:27, spiro-trikaliotis<notifications@github.com> wrote:   
While discussing all these changes, the question is: Has this ever been benchmarked? Yes, I argued with the extra cycle in many cases above, @greg-king5 pointed out incsp2 (which I had seen myself, too).
However, I am not quite sure how often these functions are actually used (that is, not how many static calls we have, but how often they are used dynamically - that can be different!), and if it is really worth the effort.

Are we speaking about 1 cycle in 1.000.000 cycles, 1 cycle in 10.000 cycles, or 1 cycle in 50 cycles here?

Does anyone know of some program (non-trivial, but deterministic) that could be used for such a benchmark?

**Comment by prandeamus** [comment]
I've prototyped a solution for cycle timing within the sim6502 application which would allow benchmarking individual chunks of code with cycle timing as accurate as those provided by the emulated CPU in the simulator. I can put this up for discussion as a pull request if there is sufficient interest. I can prove the concept but I'm not sure the design is aligned with the rest of the project's coding styles and approaches.

Let me know if anyone wants to take a look at the work in progress. I'm happy to collaborate but I don't want to create for pull requets for things that no one cares about except me!

**Comment by prandeamus** [comment]
After a cycle timer (as above) it would be good to have code coverage statistics, but that's asking for quite a lot at this point.

**Comment by oliverschmidt** [comment]
> Let me know if anyone wants to take a look at the work in progress.

I personally don't have the resources to do so.

**Comment by laubzega** [comment]
Hey, what happened to this? Having a profiler in cc65 would be cool.

**Comment by prandeamus** [comment]
Truthfully, I had other things to do ... :( I think I have a prototype profiler within sim6502 somewhere, but I lack to the time to properly evaluate it or indeed finish it. I am happy to share with other interested parties if anyone wants to complete it or whatever.
